Add Assist OSS docs (#26462)
gravitational/teleport: docs/pages/ai-assist.mdx

---
title: Try out Teleport Assist
description: This tutorial will guide you through the steps needed to run Teleport Assist on your own infrastructure.
---

<Admonition type="tip" title="Preview">
  Teleport Assist it currently in Preview.
</Admonition>

This guide will help you understand how to set up and use Teleport Assist, an AI-powered assistant that helps you run
commands, debug and navigate through your infrastructure.
<Details
  title="Version warning"
  opened={true}
  min="12.0"
>
Teleport Assist is available starting from Teleport `v12.4`.
</Details>

![Teleport Assist Example](../img/assist/assist-example.png)

## Prerequisites

Before you get started with Teleport Assist, make sure you have the following:

> (!docs/pages/includes/edition-prereqs-tabs.mdx!)

Author: Enterprise still applies here. You can still configure Ent if you really want to.

- **OpenAI Account**: You will need an active OpenAI account with GPT-4 API access as Teleport Assist relies on OpenAI services.

<Notice type="note">
If you don't have access to OpenAI GPT-4 API, you can still try out Teleport Assist by signing up for a [Teleport Team plan](https://goteleport.com/pricing/).
</Notice>

Teleport Assist currently doesn't support:
* Etcd backend
* per-session MFA (support coming soon)

If your installation is using Etcd or requires per-session MFA,
you can create a new cluster with any other backend without per-session MFA or sign up for Teleport Cloud Team plan.

## Step 1/3. Generate an OpenAI API key

Teleport Assist uses the GPT-4 model. Your account must have access to it if you want to use Teleport Assist.

Follow these steps to generate your OpenAI API key:

1. Sign in to your OpenAI account or [sign up](https://platform.openai.com/signup) if you don't have one.
2. Navigate to the [API section](https://platform.openai.com/account/api-keys) in your OpenAI dashboard.
3. Click on "Create new secret key".
4. Give your key a descriptive name.
5. Click "Create secret key".

> 6. Your new API key will be displayed. Make sure to copy it and save it in a file on the system where Teleport Proxy is running, e.g., `/etc/teleport/openai_key`.

Reviewer: ```suggestion
6. Your new API key will be displayed. Make sure to copy it and save it in a file on the system where the Teleport Proxy Service is running, e.g., `/etc/teleport/openai_key`.
```

If you have multiple instances of Teleport Proxy, you must copy the file to all of them.

> 7. Set read-only permissions and change the file owner to the user that Teleport Proxy uses ex. `teleport` by running the following commands:

Author: Proxy, even when running in systemd, should not run as root. I'm not sure how can I refer to an arbitrary user, when most people just want to copy and paste the command.

---
title: Try out Teleport Assist.
description: This tutorial will guide you through the steps needed to run Teleport Assist on your own infrastructure.
---

<Admonition type="tip" title="Preview">
  Teleport Assist it currently in Preview.
</Admonition>

> This guide will help you understand how to set up and use Teleport Assist, a new feature that enhances your remote access experience.

Reviewer 2: Replace "enhances your remote access experience" with what it actually allows you to do, otherwise reads like fluff.

> 

Author: @ptgott How/where can I create one?

> 

Author: Added

> 

Reviewer: I'll answer this outside this thread.

I think that if this guide is descriptive about the architecture of the feature, we can also add a diagram in a separate PR. 

We should definitely add a short paragraph about how the feature runs to the start of the guide.

> 

Author: Added

## Prerequisites

Before you get started with Teleport Assist, make sure you have the following:

1. **Teleport**: Ensure you have Teleport 12.4 or newer installed and running on your system.
You can check out the [Teleport Installation Guide](https://goteleport.com/docs/installation/) for more details.

2. **OpenAI Account**: You will need an active OpenAI account with GPT-4 API access as Teleport Assist relies on OpenAI services.

<Admonition
  type="tip"
  title="Cloud Team plan"
>

> If you don't have access to OpenAI GPT-4 API, you can still try out Teleport Assist by signing up for a [Teleport Team plan](https://goteleport.com/pricing/).

Reviewer: In that case, we should change the `forScopes` setting for this page in `docs/config.json` to remove the `cloud` and `enterprise` options. We'll add a separate `team` scope in a later PR.

</Admonition>

> Teleport Assist doesn't support Etcd backend or per-session MFA. If your installation is using Etcd or requires per-session MFA,

Author: >Why doesn't it support Per-Session MFA?

The support will be added soon. I can mention that.

>How does Teleport Assist use backends?

@ptgott It feels like the wrong place to mention that backend is not supported, but we're using others for.... Should we just mention how backedns are used in the architecture section?

with any other backend or sign up for Teleport Cloud Team plan.

## Step 1/3. Generate OpenAI API Key

> Teleport uses the GPT-4 model for the Assist feature. Your account must have access to it if you want to use Teleport Assist.

Reviewer 2: ```suggestion
Teleport Assist uses the GPT-4 model. Your account must have access to it if you want to use Teleport Assist.
```

Follow these steps to generate your OpenAI API key:

1. Sign in to your OpenAI account.
2. Navigate to the [API section](https://platform.openai.com/account/api-keys) in your OpenAI dashboard.
3. Click on "Create new secret key".
4. Give your key a descriptive name.
5. Click "Create secret key".

> 6. Your new API key will be displayed. Make sure to copy it and save it in a file on your system, ex. `/etc/teleport/openai_key`.

Author: The key should be added to ALL Teleport Proxy instances. I added more details to this point.

7. Set read only permissions and change the file owner to the user that Teleport Proxy uses ex. `teleport` by running the following commands:

```bash
chmod 400 /etc/teleport/openai_key
chown teleport:teleport /etc/teleport/openai_key
```

> Remember, your API keys carry many privileges, so be sure to keep them secure! Do not share your secret API keys in publicly accessible areas.

Reviewer 2: Nit: Make this into a warning admonition?

Author: Added `Notice`

## Step 2/3. Enable Assist

> To enable Teleport Assist, you need to provide your OpenAI API key. Here's how to do it:

Reviewer 2: ```suggestion
To enable Teleport Assist, you need to provide your OpenAI API key:
```

1. Open your Teleport configuration file. This is typically located at `/etc/teleport.yaml`.

> 2. Locate the `assist` section in the file. If it doesn't exist, create it in a Teleport Proxy section.

Reviewer 2: I think step 2 here can be removed. The config snippet makes it clear enough.

3. Add your OpenAI API key to the `assist` section:

 ```yaml
proxy_service:
  assist:
    openai:
      api_token_path: /etc/teleport/openai_key
```

4. Save the changes and close the file.
5. Restart your Teleport Proxy service for the changes to take effect.

> You have now enabled Teleport Assist!

Reviewer 2: ```suggestion
```

> Make sure that your Teleport user has the `assist` permission. By default, users with `auditor` and `editor` roles

Reviewer 2: ```suggestion
Make sure that your Teleport user has the `assist` permission. By default, users with built-in `auditor` and `editor` roles
```

> have this permission. You can also add it to a custom role. For more details, check out the [Create a custom role](https://goteleport.com/docs/access-controls/getting-started/#step-33-create-a-custom-role).

Author: Added

> ## Step 3/3. Example Use

Reviewer: ```suggestion
## Step 3/3. Example use
```
We use sentence case in headings

> Now that you have Teleport Assist enabled, you can start using it to enhance your remote access experience.
> Here's an example of how to use it:

Reviewer 2: ```suggestion
Now that you have Teleport Assist enabled, you can start using it.
```

1. Open Teleport's Web UI in your browser (e.g., https://teleport.example.com).

> 2. Login to the Web UI using your Teleport credentials.

Reviewer: ```suggestion
2. Log in to the Web UI using your Teleport credentials.
```
"Login" is a noun

3. Click on the "Assist" button in the top right context menu.
![Teleport Assist Menu](../img/assist/main-screen.png)
4. Click "New Conversation" to start a new conversation with Teleport Assist.
![Teleport Assist New Chat](../img/assist/assist-screen.png)

Remember, Teleport Assist is powered by OpenAI, so the more specific your query, the better the response will be.

And that's it! You have now learned how to set up and use Teleport Assist. For more details and advanced configuration options, check out the other pages in our documentation.

> If you encounter any issues, don't hesitate to reach out to our support team. Happy teleporting!

Reviewer 2: Instead of this, I would add a "Next steps" section with a few links to other guides that may be useful. E.g. joining servers or configuring RBAC. Like other guides have.
